[16.0][IMP] mail_gateway_whatsapp: Add support for WhatsApp templates

carlos-lopez-tecnativa opened the pull request:
If the company wants to initiate a conversation with a customer, a template must be used; otherwise, messages will not be sent. This also applies when the last conversation with the customer is older than 24 hours.

- Enabled downloading templates from META
- Added the ability to create templates directly in Odoo.

Note: Currently does not support templates with variables or buttons

TT51660
@Tecnativa @pedrobaeza @chienandalu @etobella could you please review this

carlos-lopez-tecnativa committed:
[IMP] mail_gateway_whatsapp: Add support for WhatsApp templates
If the company wants to initiate a conversation with a customer, a template must be used; otherwise, messages will not be sent.
This also applies when the last conversation with the customer is older than 24 hours.

- Enabled downloading templates from META
- Added the ability to create templates directly in Odoo.

Note: Currently, templates with variables or buttons are not supported.

matteoopenf commented:
Hi, do you plan to do a backup for the module to 14.0?

pedrobaeza commented:
Definitely no.

carlos-lopez-tecnativa commented:
I fixed the test issues related to the module `mail_gateway_whatsapp`, but I'm not sure if the other errors are related(`mail_composer_cc_bcc_account`). Could you please take a look? @pedrobaeza @etobella
https://github.com/OCA/social/actions/runs/11842395654/job/33000889279?pr=1497#step:8:1146

carlos-lopez-tecnativa committed:
[IMP] mail_gateway: Fine-tune default values for gateway behavior related to the current user

- Use OdooBot for webhooks instead of the current user, following the same logic as Odoo.
- Avoid automatically adding the current user as a member of a new gateway to prevent notifications for all messages. This is now an explicit configuration.

etobella reviewed (approved):
At a technical Level LGTM

etobella commented on mail_gateway/models/mail_gateway.py:
```diff
@@ -25,12 +25,10 @@ class MailGateway(models.Model):
         help="User that will create the messages",
     )
-    member_ids = fields.Many2many(
-        "res.users", default=lambda self: [Command.link(self.env.user.id)]
```
I think this changes are the reason behind the test issue

carlos-lopez-tecnativa replied:
> I think this changes are the reason behind the test issue

The CI has been failing for several days on all PRs/commits in V16.

The last commit is unrelated to WhatsApp templates. If you prefer, I can create a new PR for it separately. However, this commit is not linked to the tests but addresses a different case, please refer to the commit description for more details.

Let me know what you think.

cuongnmtm reviewed (approved):
LGTM (code review and functional test).

etobella commented:
CI is working in main branch, can you apply my suggestion and rebase?

carlos-lopez-tecnativa commented:
> CI is working in main branch, can you apply my suggestion and rebase?

I have updated the code and fixed the test. Could you please review it again?

etobella commented on mail_gateway/models/mail_gateway.py:
```diff
@@ -25,12 +25,10 @@ class MailGateway(models.Model):
     webhook_user_id = fields.Many2one(
         "res.users",
-        default=lambda self: self.env.user.id,
+        default=lambda self: self.env.ref("base.user_root"),
```
I don't understand why do you change this. What is the problem of the current default of these fields?

carlos-lopez-tecnativa replied:
If the settings are configured by a regular user who does not wish to receive system notifications for every message, it is better not to add them automatically. Additionally, for the webhook user, Odoo follows a similar logic by using the Odoo bot user instead of the current user.

pedrobaeza replied:
@etobella can you please check?

etobella replied:
@pedrobaeza I reviewed and it looked good for me, I just made a small suggestion, but we can continue with merge.

pedrobaeza reviewed (approved):
/ocabot merge minor
